### ISSUE TITLE
JSONSerializer Numbers default to Int64

### ISSUE THREAD

**Issue by rofr** [DevrexLabs/memstate#19]
```csharp
        [Fact]
        public void Int32_can_be_cloned()
        {
            int i = 42;
            var serializer = new JsonSerializerAdapter();
            var stream = new MemoryStream();
            serializer.WriteObject(stream, i);
            stream.Position = 0;
            object o = serializer.ReadObject(stream);
            Assert.IsType<int>(o); //fails, o is long
        }

        [Fact]
        public void Int64_can_be_cloned()
        {
            long i = 42;
            var serializer = new JsonSerializerAdapter();
            var stream = new MemoryStream();
            serializer.WriteObject(stream, i);
            stream.Position = 0;
            object o = serializer.ReadObject(stream);
            Assert.IsType<long>(o); //succeeds
        }
```

**Comment by baskar3078** [comment]
Am interested to contribute , let me know how I can help.

**Comment by rofr** [comment]
sounds great! Do you mean with this issue or in general?

The problem here is that the newtonsoft json serializer will deserialize an integer (presumably also byte and short) as a long when type information is missing. I see  2 approaches to dealing with this:

1. document the behavior so that users can avoid sending untyped numbers
2. Create a workaround in the adapter

So if you can figure out a good workaround that is transparent to the user that would be great. More specifically, the 2 test cases above should pass, not just the second one.

**Comment by baskar3078** [comment]
I reached out to contribute in general and help with this issue as well. let me know the files where changes are required. I will send a fork request and make the required changes.

**Comment by rofr** [comment]
The tests above are in Memstate.Tests/SerializationTests.cs
The serializer itself is a wrapper around NewtonSoft.JsonSerializer, it lives in the Memstate.JsonNet project.
If you want to discuss an approach, have any questions or just want to hang out, find us in the gitter.im/memstate channel

**Comment by baskar3078** [comment]
What are the use cases for serializer adapter. If it is not going to be called from other projects then we can include a xml documentation for the method to always pass signed type and by default handle any int variables as long.

**Comment by rofr** [comment]
JsonSerializerAdapter implements ISerializer. ISerializers are used for multiple purposes: 

- cloning object graphs to ensure isolation, results passed back from queries/commands
- serializing/deserializing commands and queries transfered from memstate client to memstate server
- serializing/deserializing commands to the command journal

Commands, queries and query results are defined by the developers using memstate. They will not interact with the adapter itself but objects they define will be cloned by the serializer. Therefore we want the serializers to be "true" (the deserialized graph must be identical, or at least semantically equivalent to the input graph).

Also, users should be able to switch between different serializers without affecting the behavior of their system. That could turn in to a mess if the developer needs to put workarounds in their code for every possible serializer.

**Comment by baskar3078** [comment]
My solution approach would be as below. This is not the clean way.
1. In the writeObject method, we can check for the type of incoming object and if it is equal to int, then we can convert it to byte array and create a base64string and perform serialization as base64string.
2. In the readObject method, we can reparse the base64string to byte array.  Below is the code for both methods which makes the test cases pass. 
public void WriteObject(Stream serializationStream, object @object)
        {
            var streamWriter = new StreamWriter(serializationStream);
            var writer = new JsonTextWriter(streamWriter);
            if (@object.GetType().ToString().Equals("System.Int32") || @object.GetType().ToString().Equals("System.Int16"))
            {
                byte[] result = new byte[1];
                result[0] = System.Convert.ToByte(@object);
                string convertedstring = System.Convert.ToBase64String(result);
                _serializer.Serialize(writer, convertedstring);
            }
            else
            {
                _serializer.Serialize(writer, @object);
            }
            writer.Flush();
            streamWriter.Flush();
        }

public object ReadObject(Stream serializationStream)
        {
            TextReader streamReader = new StreamReader(serializationStream);
            string line = streamReader.ReadLine();
            var reader = new JsonTextReader(new StringReader(line));
            string output = _serializer.Deserialize(reader).ToString();

            if (!string.IsNullOrEmpty(output) && output.EndsWith("=="))
            {
                try
                {
                    byte[] result = System.Convert.FromBase64String(output);
                    return Convert.ToInt32(result[0]);
                }
                catch(Exception ex)
                {
                     return Convert.ToInt64(output);
                }
            }
            else
            {
                return _serializer.Deserialize(reader);
            }
        }

**Comment by rofr** [comment]
This "solves" the case when an int or short is passed directly to the Serialize method. Another case is that an int or short stored in a property of type Object will be deserialized as a long, making the following fail:
```csharp
class Foo
{
   public object Number { get; set;]
}

int number = 42;
var foo = new Foo{ Number = number};
var fooCopy = serializer.Clone(foo);
Assert.IsType<int>(fooCopy.Number);
```

**Comment by rofr** [comment]
Your approach converts/deconverts numbers in a transparent manner to some other type avoiding the bug. This is a good idea but I think there is a better intermediate format than base64, a custom wrapper for example: 
```csharp
class Int32Wrapper
{
   public int Value { get; set; }
}
```

**Comment by baskar3078** [comment]
Yes JsonConverter is the way to go as I initial thought. I will try to implement using JsonConverter.

**Comment by rofr** [comment]
@baskar3078  great! @rogeralsing shared links on gitter to some almost identical work, see https://gitter.im/memstate/Lobby

**Comment by baskar3078** [comment]
I will be working on this over the weekend and let you know the updates.

**Comment by rofr** [comment]
@baskar3078 great! Find us in the gitter chat if you want to discuss anything or just hang out :)

**Comment by baskar3078** [comment]
[JsonSerializerAdapter.txt](https://github.com/DevrexLabs/memstate/files/1516560/JsonSerializerAdapter.txt)
[SurrogateConverter.txt](https://github.com/DevrexLabs/memstate/files/1516561/SurrogateConverter.txt)

Let me know if this changes are fine.

**Comment by rofr** [comment]
Looks great, nice work! The normal procedure for contributions is that you fork the repository and then create a Pull Request (PR) based on a commit in your fork. This is a common practice across all of GH.

**Comment by rofr** [comment]
Fixed by #35
